README.md
# Transfer of entities

Read this in another language | [English](/README.md) | [Русский](/docs/ru/README.md)
|---|---|---|

## Quick Links

[Changelog](CHANGELOG.md) | [Contributing](CONTRIBUTING.md)
|---|---|

## Contents

* [Overview](#overview)
* [Issues](#issue)
* [Features](#feature)
* [Installing](#installing)
* [License](#license)

## Overview

Allows to transfer entities to other teams.

## <a name="issue"></a> Found an Issue?

Please report any issues or a mistake in the documentation, you can help us by [submitting an issue][issues] to our GitLab Repository or on [mods.factorio.com][mod portal] or on [forums.factorio.com][homepage].

## <a name="feature"></a> Want a Feature?

You can *request* a new feature by [submitting an issue][issues] to our GitLab Repository or on [mods.factorio.com][mod portal] or on [forums.factorio.com][homepage].

## Installing

If you have downloaded a zip archive:

* simply place it in your mods directory.

For more information, see [Installing Mods on the Factorio wiki](https://wiki.factorio.com/index.php?title=Installing_Mods).

If you have downloaded the source archive (GitLab):

* copy the mod directory into your factorio mods directory
* rename the mod directory to transfer_entities_*versionnumber*, where *versionnumber* is the version of the mod that you've downloaded (e.g., 1.0.1)

## License

This project is copyright © 2019 ZwerOxotnik \<zweroxotnik@gmail.com\>.

Use of the source code included here is governed by the MIT licence. See the [LICENCE](/LICENCE) file for details.

[issues]: https://gitlab.com/ZwerOxotnik/transfer_entities/issues
[mod portal]: https://mods.factorio.com/mod/transfer_entities/discussion
[homepage]: https://forums.factorio.com/viewtopic.php?f=190&t=67245
[Factorio]: https://factorio.com/


CHANGELOG.md
# Changelog

## 2019-04-22

### [v1.0.1][v1.0.1]

- Fixed loading in special cases

## 2019-03-09

### [v1.0.0][v1.0.0]

- First release for Factorio 0.17

[v1.0.1]: https://mods.factorio.com/mod/transfer_entities/downloads
[v1.0.0]: https://mods.factorio.com/mod/transfer_entities/downloads


data.lua
-- Copyright (c) 2019 ZwerOxotnik <zweroxotnik@gmail.com>
-- Licensed under the MIT licence;

local util = require("data/util/te_util")
local tools = require("shared").transfer_entities_tools
local path = util.path("data/transfer_entities/")

local selection_mode = {"same-force", "entity-with-health"}
local transfer_entities_tool =
{
	type = "selection-tool",
	name = tools.transfer_entities_tool,
	localised_name = {"item-name." .. tools.transfer_entities_tool},
	selection_mode = selection_mode,
	alt_selection_mode = selection_mode,
	selection_cursor_box_type = "copy",
	alt_selection_cursor_box_type = "pair",
	icon = path .. "transfer_entities.png",
	icon_size = 128,
	stack_size = 1,
	flags = {},
	show_in_library = true,
	selection_color = {g = 1},
	alt_selection_color = {g = 1, b = 1},
	draw_label_for_cursor_render = true,
}
local select_units_shortcut =
{
	type = "shortcut",
	name = tools.transfer_entities_shortcut,
	order = "y",
	action = "create-blueprint-item",
	localised_name = {"item-name." .. tools.transfer_entities_tool},
	item_to_create = tools.transfer_entities_tool,
	style = "blue",
	icon =
	{
		filename = path .. "transfer_entities_shortcut.png",
		size = 128,
		priority = "extra-high-no-scale",
		flags = {"icon"}
	}
}

data:extend{
	transfer_entities_tool,
	select_units_shortcut
}


transfer_entities/gui.lua
-- Copyright (c) 2019 ZwerOxotnik <zweroxotnik@gmail.com>
-- Licensed under the MIT licence;

local module = {}

module.destroy_gui = function(player)
    local frame = player.gui.left.transfer_entities_frame
    if not frame then return end
    frame.destroy()
end

module.create_gui = function(player)
    -- Find forces
    local forces = {}
    local is_check_online = settings.global["TE-only_online_force"].value
    local player_force = player.force
    if settings.global["TE-to_whom"].value == "ally" then
        for _, force in pairs(game.forces) do
            if not is_check_online or #force.connected_players > 0 then
                if player_force.get_friend(force) and player.force ~= force then -- Is ally?
                    table.insert(forces, force.name)
                end
            end
        end
    else
        for _, force in pairs(game.forces) do
            if (not is_check_online or #force.connected_players > 0) and player.force ~= force then
                if player_force.get_friend(force) then  -- Is ally?
                    table.insert(forces, force.name)
                elseif player_force.get_cease_fire(force) then  -- Is neutral?
                    table.insert(forces, force.name)
                end
            end
        end
    end
    if #forces == 0 then
        local message = {"transfer_entities.forces-not-found"}
        local color = {1, 1, 0}
        local character = player.character
        if character then
            rendering.draw_text{
                text = message,
                scale = 0.85,
                surface = character.surface,
                target = character,
                target_offset = {0.5, -1.5},
                color = {1, 1, 0},
                time_to_live = 230,
                players = {game.player},
                alignment = "left",
                scale_with_zoom = true
            }
        else
            player.print(message, color)
        end
        module.destroy_gui(player)
        global.transfer_entities.players[player.index].transfer_entities = nil
        return
    end

    module.destroy_gui(player)
    -- Create gui
    local left = player.gui.left
    local frame = left.add{type = 'frame', name = 'transfer_entities_frame', caption = {"mod-name.transfer_entities"}}
    local main_table = frame.add{type = 'table', name = 'main_table', column_count = 3}
    local x = main_table.add{type = 'button', name = "X", caption = 'X', tooltip = {"controls.close-gui"}}
    x.style.minimal_width = 30
    x.style.maximal_width = 30
    x.style.font = 'default'
    x.style.left_padding = 0
    x.style.top_padding = 0
    x.style.bottom_padding = 0
    x.style.right_padding = 0
    local drop_down_forces = main_table.add{type = 'drop-down', name = 'to_force', items = forces, selected_index = 1}
    drop_down_forces.style.maximal_width = 240
    local button = main_table.add{type = 'button', name = "transfer_entities", caption = '➤', tooltip = {"gui-train-rename.perform-change"}}
    button.style.maximal_height = 22
    button.style.minimal_height = 22
    button.style.minimal_width = 24
    button.style.maximal_width = 24
    button.style.font = 'default'
    button.style.left_padding = 0
    button.style.top_padding = 0
    button.style.bottom_padding = 0
    button.style.right_padding = 0
end

module.on_gui_click = function(event)
    -- Validation of data
	local gui = event.element
	if not (gui and gui.valid) then return end
	local player = game.players[event.player_index]
	if not (player and player.valid) then return end
	local parent = gui.parent
    if not parent then return end
    if parent.name ~= "main_table" then return end

    if gui.name == "X" then
        global.transfer_entities.players[player.index].transfer_entities = nil
        module.destroy_gui(player)
    elseif gui.name == "transfer_entities" then
        local drop_down = parent.to_force
        local new_force = game.forces[drop_down.items[drop_down.selected_index]]
        if not new_force then
            local message = {"transfer_entities.forces-not-found"}
            local color = {1, 1, 0}
            local character = player.character
            if character then
                rendering.draw_text{
                    text = message,
                    scale = 0.85,
                    surface = character.surface,
                    target = character,
                    target_offset = {0.5, -1.5},
                    color = {1, 1, 0},
                    time_to_live = 230,
                    players = {game.player},
                    alignment = "left",
                    scale_with_zoom = true
                }
            else
                player.print(message, color)
            end
            return
        end

        -- Set new force
        local transfer_entities = global.transfer_entities.players[player.index].transfer_entities
        for _, entity in pairs(transfer_entities) do
            entity.force = new_force
        end
        transfer_entities = nil
        module.destroy_gui(player)
    end
end

-- module.on_gui_selection_state_changed = function(event)
--     -- Validation of data
-- 	local gui = event.element
-- 	if not (gui and gui.valid) then return end
-- 	local player = game.players[event.player_index]
-- 	if not (player and player.valid) then return end
-- 	local parent = gui.parent
-- 	if not parent then return end
-- 	if not (parent.name == "to_force" and parent.name == "main_table") then return end
-- end

return module


transfer_entities/control.lua
--[[

Copyright (c) 2019 ZwerOxotnik <zweroxotnik@gmail.com>
Licensed under the MIT licence;
Author: ZwerOxotnik
Version: 1.0.1 (2019.04.22)

Description: allows to transfer entities to other teams

You can write and receive any information on the links below.
Source: https://gitlab.com/ZwerOxotnik/transfer_entities
Mod portal: https://mods.factorio.com/mod/transfer_entities
Homepage: https://forums.factorio.com/viewtopic.php?f=190&t=67245

]]--

local transfer_entities_tool = require("shared").transfer_entities_tools.transfer_entities_tool
local data =
{
	players = {}
}

local gui = require("transfer_entities/gui")
local module = {}
module.events = {}
module.version = "1.0.1"

local get_event
if event_listener then
	get_event = function(event)
		return defines.events[event] or event
	end
else
	get_event = function(event)
		if type(event) == "number" then
			return event
		else
			return defines.events[event]
		end
	end
end

-- This function for compatibility with "Event listener" module and into other modules
local function put_event(event, func)
	event = get_event(event)
	if event then
		module.events[event] = func
		return true
	else
		log("That event is nil")
		-- error("That event is nil")
	end
	return false
end

module.on_init = function()
  	global.transfer_entities = data
end

module.on_load = function()
	data = global.transfer_entities
end

local function delete_selected_entities(event)
	if not data.players[event.player_index] then return end
	data.players[event.player_index].transfer_entities = nil
	local player = game.players[event.player_index]
	if not (player and player.valid) then return end
	gui.destroy_gui(player)
end

local function check_selected_enities(event)
	-- Validation of data
	local item = event.item
	if item ~= transfer_entities_tool then return end
	local player = game.players[event.player_index]
	if not (player and player.valid) then return end
	local entities = event.entities
	if not entities then return end

	local max_entities = settings.global["TE-max_entities"].value
	if not player.admin and #entities > max_entities then
		local message = {"transfer_entities.too-much-entities", max_entities}
		local color = {1, 1, 0}
		local character = player.character
		if character then
			rendering.draw_text{
				text = message,
				scale = 0.85,
				surface = character.surface,
				target = character,
				target_offset = {0.5, -1.5},
				color = {1, 1, 0},
				time_to_live = 230,
				players = {game.player},
				alignment = "left",
				scale_with_zoom = true
			}
		else
			player.print(message, color)
		end
		return
	end

	data.players[event.player_index].transfer_entities = {}
	local transfer_entities = data.players[event.player_index].transfer_entities
	for entity_number, entity in pairs(entities) do
		if entity.valid and entity.type ~= "player" then
			transfer_entities[entity_number] = entity
		end
	end

	if #transfer_entities == 0 then
		-- is print?
		return
	end

	gui.create_gui(player)
end

local function delete_player_data(event)
	data.players[event.player_index] = nil
end

local function on_player_created(event)
	data.players[event.player_index] = {}
end

local function on_player_joined_game(event)
	-- Validation of data
	local player = game.players[event.player_index]
	if not (player and player.valid) then return end

	data.players[event.player_index] = {}
end

local function on_runtime_mod_setting_changed(event)
	if event.setting_type ~= "runtime-global" then return end

	local event_name = event.setting
	if event_name == "TE-only_online_force" or event_name == "TE-to_whom" then
		for player_index, player in pairs(game.connected_players) do
			data.players[player_index].transfer_entities = nil
			gui.destroy_gui(player)
		end
	end
end

-- put_event("on_configuration_changed", on_configuration_changed)
put_event("on_gui_click", gui.on_gui_click)
-- put_event("on_gui_selection_state_changed", gui.on_gui_selection_state_changed)
put_event("on_player_removed", delete_player_data)
put_event("on_player_died", delete_selected_entities)
put_event("on_player_created", on_player_created)
put_event("on_player_joined_game", on_player_joined_game)
put_event("on_player_left_game", delete_selected_entities)
put_event("on_player_changed_force", delete_selected_entities)
put_event("on_player_changed_surface", delete_selected_entities)
put_event("on_player_selected_area", check_selected_enities)
put_event("on_runtime_mod_setting_changed", on_runtime_mod_setting_changed)

return module
